### ISSUE TITLE
Sentence deletion segfaults

### ISSUE THREAD

**Issue by 0branch** [lichray/nvi2#33]
(Encountered under OS X.)

First case:

- `printf "\nfoo" > a && nvi a`
- `d(`

Second case:

- `printf "bar\n\n  baz" > b && nvi +3 b`
- `d(`

Haven't had time to look into this yet; here's the offending call,

    (lldb) up
    frame #2: 0x000000010000b0b1 nvi`del(sp=0x0000000100805400, fm=0x00007fff5fbff2a8, tm=0x00007fff5fbff2b8, lmode=0) + 1969 at delete.c:100
       97  			GET_SPACE_RETW(sp, bp, blen, len);
       98  			if (fm->cno != 0)
       99  				MEMCPY(bp, p, fm->cno);
    -> 100 			MEMCPY(bp + fm->cno, p + (tm->cno + 1),
       101 				len - (tm->cno + 1));
       102 			if (db_set(sp, fm->lno,
       103 			    bp, len - ((tm->cno - fm->cno) + 1)))

Frame variables (`break set -f delete.c -l 100`):

    (lldb) fr var
    (SCR *) sp = 0x0000000100805400
    (MARK *) fm = 0x00007fff5fbff288
    (MARK *) tm = 0x00007fff5fbff298
    (int) lmode = 0
    (recno_t) lno = 2
    (size_t) blen = 256
    (size_t) len = 0
    (size_t) nlen = 210453397506
    (size_t) tlen = 0
    (CHAR_T *) bp = 0x000000010020b1a0 L"\U0000fffd\U0000fffd\U0000fffd\U0000fffd\U0000fffd\U0000fffd\U0000fffd\U0000fffd਩": unmodified: line 1 (pid 28160)
    (CHAR_T *) p = 0x000000010020b2f0 L"foo"
    (int) eof = 1
    (int) rval = 8410112

**Comment by 0branch** [comment]
It turns out that both cases segfault vanilla 1.79 (running under Linux).

**Comment by lichray** [comment]
This patch https://github.com/lichray/nvi2/commit/370e056817f3b9c0adc5ee553553dda5fa172121 should "fix" the issue.  This piece of code tries to quit sentence detection too early.  But don't know how to get this part done right, yet.

**Comment by lichray** [comment]
Zzz, completely wrong.

**Comment by 0branch** [comment]
There actually seem to be two separate issues:

1. `(` can move the cursor forwards. This was fixed in Debian a while ago (see [here](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=193498)). I'm pretty sure that this is a bug in nvi and not for compatability ("[traditional](http://ex-vi.sourceforge.net/)" vi moves backwards).
2. When deletion within a single line is triggered by `d(`, `len` is 0 if no movement occurred (unlike `k` and friends, repeated `(`s at SOF are silently permitted). Since there's no error, `MEMCPY` is passed a size of -1 (`len - (tm->cno + 1)`) and segfaults.

Draft patch [here](https://github.com/0branch/nvi2/commit/fa513e).

**Comment by lichray** [comment]
Your analysis is right.  Just I was a little bit unsure whether Debian's fix is at the right place.  `vi_sentenceb` has the same condition, but runs in a different flow.

**Comment by 0branch** [comment]
>  `vi_sentenceb` has the same condition, but runs in a different flow.

Are you referring to [this](https://github.com/lichray/nvi2/blob/master/vi/v_sentence.c#L79-80) condition in `vi_sentencef` (rather than  `vi_sentenceb`)?

**Comment by lichray** [comment]
Yes.

**Comment by 0branch** [comment]
Sorry for the delay. Thoughts:

- The problematic behaviour in `(` is due to a compensatory forward movement (through blanks) yielding a cursor position after the intended column. 
- In `)`, we only attempt forward movement through whitespace; this is correct and anticipated. As such, we're just checking whether movement occurred at all (should the count hit zero).
- The condition in `vi_sentencef` could use a different comparison operator to more directly reflect directionality, though this shouldn't be necessary.
- The change in operator for `vi_sentenceb` is required due to blind application of `cs_next`/`cs_fblank`. We not only need to check whether movement occured, but also whether the resultant direction is appropriate.

In other words, the difference lies in the use of forward-eating whitespace functions: defensive direction checks aren't required for `vi_sentencef`.

**Comment by 0branch** [comment]
(Bump.)

**Comment by lichray** [comment]
I know... I'm traveling without my BSD laptop.  I will take this 2 weeks later.
